### PR TITLE
refactor[next]: make function definition -> PAST into a workflow step

### PR DESCRIPTION
**Pull request by DropD**
### New:

- `recipes.ProgramTransformWorkflow`, a transform workflow with support for partial transforms by giving access to steps.
- `ffront.func_to_past.func_to_past`, a transform workflow step `stages.ProgramDefinition` -> `stages.ProgramPast`
- workflow stages:
  - `ProgramDefinition`: function object + grid type
  - `ProgramPast`: PAST node + closure vars + grid type

### Changed:

- `backend.Backend`:
  - `.transformer` is now specifically expected to be a `recipes.ProgramTransformWorkflow`, to allow partial transformations 
- `decorator.Program`:
  - data attributes: packed into `.definition_stage`, except for `.backend`.
  - properties: `.definition`, `.grid_type`, `.past_node`, `.closure_vars` provide access to the same data as before
  - new property: `.past_stage` (cached) does (argument-less) function -> PAST lowering on demand, using the backend's transform workflow if it exists.

## Requirements

- [ ] All fixes and/or new features come with corresponding tests.
- [ ] Important design decisions have been documented in the approriate ADR inside the [docs/development/ADRs/](docs/development/ADRs/Index.md) folder.

If this PR contains code authored by new contributors please make sure:

- [ ] All the authors are covered by a valid contributor assignment agreement provided to ETH Zurich and signed by the employer if needed.
- [ ] The PR contains an updated version of the `AUTHORS.md` file adding the names of all the new contributors.

### CODE REVIEW

File: src/gt4py/next/ffront/decorator.py [hunk 1]
```diff
@@ -124,18 +120,34 @@ def from_function(
...
-            grid_type=grid_type,
         )
 
+    @property
+    def definition(self):
+        return self.definition_stage.definition
+
+    @property
+    def grid_type(self):
+        return self.definition_stage.grid_type
```
Comment (havogt) [inline src/gt4py/next/ffront/decorator.py:135]: I tend to not expose this here (if it's needed during refactoring, I'd add a TODO to remove it later)

File: src/gt4py/next/ffront/decorator.py [hunk 2]
```diff
@@ -124,18 +120,34 @@ def from_function(
...
+            return self.backend.transformer.func_to_past(self.definition_stage)
+        return next_backend.DEFAULT_TRANSFORMS.func_to_past(self.definition_stage)
+
+    @property
+    def past_node(self):
+        return self.past_stage.past_node
+
+    @property
+    def closure_vars(self):
+        return self.past_stage.closure_vars
```
Comment (havogt) [inline src/gt4py/next/ffront/decorator.py:149]: Same, probably not needed.

**Review by havogt — COMMENTED** [review]
I think it looks good. Will have a closer look tomorrow. Maybe we should think of a simple visualization tool. It will get tricky to follow in code how the workflow is composed.

File: src/gt4py/next/ffront/decorator.py (hunk 2 above)
Comment (DropD) [reply src/gt4py/next/ffront/decorator.py:149]: Some of this stuff might be used in icon4py. I would suggest adding a todo comment and a deprecation warning.

File: src/gt4py/next/otf/stages.py [hunk 3]
```diff
@@ -32,11 +33,24 @@
...
 
 
+@dataclasses.dataclass(frozen=True)
+class ProgramDefinition:
+    definition: types.FunctionType
+    grid_type: Optional[common.GridType] = None
+
+
+@dataclasses.dataclass(frozen=True)
+class ProgramPast:
```
Comment (havogt) [inline src/gt4py/next/otf/stages.py:43]: I don't know of a really good name, but maybe `PastProgramDefinition`?

**Review by havogt — APPROVED** [review]
2 comments, but approved. Your decision what to do.

File: src/gt4py/next/otf/stages.py [hunk 4]
```diff
@@ -32,11 +33,24 @@
 SettingT_co = TypeVar("SettingT_co", bound=languages.LanguageSettings, covariant=True)
 
 
+@dataclasses.dataclass(frozen=True)
```
[the remark this answers is not in the document]
Comment (DropD) [reply src/gt4py/next/otf/stages.py:33]: I'll move these to `ffront.stages` then.

File: src/gt4py/next/otf/stages.py (hunk 3 above)
Comment (DropD) [reply src/gt4py/next/otf/stages.py]: 👍

**Pull request merged**